xdsl/transforms/experimental/stencil_global_to_local.py

```diff
@@ -0,0 +1,491 @@
+        self.core_lb = core_lb
+        self.core_ub = core_ub
+
+    @staticmethod
+    def from_halo_swap_op(op: stencil.HaloSwapOp):
+        assert op.buff_lb is not None, "HaloSwapOp must be lowered after shape inference!"
+        assert op.buff_ub is not None, "HaloSwapOp must be lowered after shape inference!"
+        assert op.core_lb is not None, "HaloSwapOp must be lowered after shape inference!"
+        assert op.core_ub is not None, "HaloSwapOp must be lowered after shape inference!"
+        return DimsHelper(
+            op.buff_lb.as_tuple(),
+            op.buff_ub.as_tuple(),
+            op.core_lb.as_tuple(),
+            op.core_ub.as_tuple(),
+        )
+
+    # positions
+
+    def buffer_start(self, dim: int):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        return self.buff_lb[dim]
+
+    def core_start(self, dim: int):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        return self.core_lb[dim]
+
+    def buffer_end(self, dim: int):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        return self.core_ub[dim]
+
+    def core_end(self, dim: int):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        return self.buff_ub[dim]
+
+    # sizes
+
+    def buff_size(self, dim: int):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        return self.buff_ub[dim] - self.buff_lb[dim]
+
+    def core_size(self, dim: int):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        return self.core_ub[dim] - self.core_lb[dim]
+
+    def halo_size(self, dim: int, at_end: bool = False):
+        assert dim < self.dims, f"The given DimsHelper only has {self.dims} dimensions"
+        if at_end:
+            return self.buff_ub[dim] - self.core_ub[dim]
+        return self.core_lb[dim] - self.buff_lb[dim]
+
+
+@dataclass
+class SlicingStrategy(ABC):
+
+    @abstractmethod
+    def calc_resize(self, shape: tuple[int]) -> tuple[int]:
+        raise NotImplementedError(
+            "SlicingStrategy must implement calc_resize!")
+
+    @abstractmethod
+    def halo_exchange_defs(self,
+                           dims: DimsHelper) -> Iterable[HaloExchangeDef]:
+        raise NotImplementedError(
+            "SlicingStrategy must implement halo_exchange_defs!")
+
+    @abstractmethod
+    def comm_count(self) -> int:
+        raise NotImplementedError("SlicingStrategy must implement comm_count!")
+
+
+@dataclass
+class HorizontalSlices2D(SlicingStrategy):
+    slices: int
+
+    def __post_init__(self):
+        assert self.slices > 1, "must slice into at least two pieces!"
+
+    def comm_count(self) -> int:
+        return self.slices
+
+    def calc_resize(self, shape: tuple[int, ...]) -> tuple[int, ...]:
+        # slice on the last dimension only
```

Comment:
last/inner

Reply:
performance wise it is better to slice the outer dimension.
So as if you split x, y
You start by splitting x